### ISSUE TITLE
Type declaration issue when no semicolon

### ISSUE THREAD

**Issue by kube** [Microsoft/TypeScript-TmLanguage#352]
<img width="671" alt="screen shot 2016-11-10 at 16 55 22" src="https://cloud.githubusercontent.com/assets/2991143/20183719/9469889c-a766-11e6-8455-8236f5e8833a.png">

<img width="692" alt="screen shot 2016-11-10 at 16 55 35" src="https://cloud.githubusercontent.com/assets/2991143/20183722/97e9b3d4-a766-11e6-83cf-4f37168dd145.png">

**[TS and JS Grammar Extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.typescript-javascript-grammar) version:**  0.4 / latest

**Code**

As you can see in attached screenshots :
`program` is not colored when `type` declaration misses the semicolon at end.

I always code without semicolons, and it's valid TypeScript.

**Comment by mhegazy** [comment]
what version of [TS and JS Grammar Extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.typescript-javascript-grammar) are you using?

**Comment by mhegazy** [comment]
can you also share a code sample we can copy/past?

**Comment by kube** [comment]
I'm using the last one from the Marketplace, I updated it today.

A sample code:

```ts
const sayHello = () => console.log('Hello')

type SomeType = {
      a: number,
      b: string
}

sayHello()
```

`sayHello()` won't be highlighted.

Add a semicolon after type declaration and it will be.

**Comment by sheetalkamat** [comment]
There is no easy way to fix this. If we try to fix this we would have issues like #166, #206 
Because there is no way to distinguish if the new line would continue the type or not we have a rule that only stops the type declaration context on semicolon or start of the new declaration statement.

**Comment by kube** [comment]
Could the end of the type declaration not be defined by the first `[^&|]` instead of `;`?

I mean semicolons are really something I don't use anymore (and that I don't like) and I don't believe to be the only one in this case. (Standard linter gets rid of them)

And coding without having syntax highlighting is not something very pleasant neither.

**Comment by sheetalkamat** [comment]
> Could the end of the type declaration not be defined by the first  [^&|]  instead of  ; ?
That does not work for scenario like 
```typescript
type A = B
 | C
```
Note that this is because grammar can only lookup one line at a time The is no way to see if there is ```|``` on new line or not.

**Comment by pirix-gh** [comment]
@sheetalkamat you can't just close an issue because it is too much work. The highlighting is incorrect. Instead, the grammar tool should be able to observe current, previous and next line to be able to compute properly (follow a type from A-to-Z) ? Could we re-open this issue ?

**Comment by RyanCavanaugh** [comment]
@pirix-gh this isn't about "too much work", it's about the fundamental design limitations of TextMate grammars (which we neither created nor control). We can't force a square peg into a round hole.

If you think we're mistaken about how the grammar works and can identify a fix, we'd welcome a PR, but our analysis of the situation is that this isn't possible to fix in a tmLanguage colorizer.

**Comment by pirix-gh** [comment]
Ok, but since this is not fixed, the issue should stay open, while waiting for someone to come up with a PR :)

**Comment by RyanCavanaugh** [comment]
As project maintainers, we decide what constitutes an "open" issue. The definition we've chosen is that issues are open if there is identifiable work to do *in this project*. Since the work is in the framework that consumes this definition, there's nothing we can do about it here, so the issue is closed.

**Comment by sheetalkamat** [comment]
This is fixed by 74f669d if type declaration follows empty line. Not having empty line will be limitation and cant be fixed.
